File: nativelib/src/main/resources/gc.c
#include <gc.h>

// At the moment we rely on the conservative
// mode of Boehm GC as our garbage collector.

void scalanative_init() {
    GC_init();
}

void* scalanative_alloc(void* info, size_t size) {
    void** alloc = (void**) GC_malloc(size);
    *alloc = info;
    return (void*) alloc;
}

File: nativelib/shared/src/main/scala/scala/scalanative/runtime/GC.scala
package scala.scalanative
package runtime

import native._

/**
 * The Boehm GC conservative garbage collector
 *
 * @see [[http://hboehm.info/gc/gcinterface.html C Interface]]
 */
@extern
object GC {
  @name("scalanative_alloc")
  def alloc(info: Ptr[ClassType], size: CSize): Ptr[Byte] = extern
  @name("scalanative_alloc_atomic")
  def alloc_atomic(info: Ptr[ClassType], size: CSize): Ptr[Byte] = extern
  @name("scalanative_collect")
  def collect(): Unit = extern
}
